src/FEValues/GeometryMapping.jl

```diff
@@ -0,0 +1,163 @@
+and the stored jacobian, `J`, and potentially hessian, `H`, are 
+used when mapping the `FunctionValues` to the current cell during `reinit!`.
+"""
+MappingValues
+
+struct MappingValues{JT, HT<:Union{Nothing,AbstractTensor{3}}}
+    J::JT # dx/dξ # Jacobian
+    H::HT # dJ/dξ # Hessian
+end
+@inline getjacobian(mv::MappingValues) = mv.J 
+@inline gethessian(mv::MappingValues{<:Any,<:AbstractTensor}) = mv.H
+
+# This will be needed for optimizing away the hessian calculation/updates
+# for cases when this is known to be zero (due to the geometric interpolation)
+#@inline gethessian(::MappingValues{JT,Nothing}) where JT = _make_hessian(JT)
+#@inline _make_hessian(::Type{Tensor{2,dim,T}}) where {dim,T} = zero(Tensor{3,dim,T})
+
+struct RequiresHessian{B} end
+RequiresHessian(B::Bool) = RequiresHessian{B}()
+function RequiresHessian(ip_fun::Interpolation, ip_geo::Interpolation)
+    # Leave ip_geo as input, because for later the hessian can also be avoided 
+    # for fully linear geometric elements (e.g. triangle and tetrahedron)
+    # This optimization is left out for now. 
+    RequiresHessian(requires_hessian(get_mapping_type(ip_fun)))
+end
+
+"""
+    GeometryMapping(::Type{T}, ip_geo, qr::QuadratureRule, ::RequiresHessian{B})
+
+Create a `GeometryMapping` object which contains the geometric shape, gradients, and, 
+if `B==true`, the hessian values. `T<:AbstractFloat` gives the numeric type of the values.
+"""
+GeometryMapping
+
+struct GeometryMapping{IP, M_t, dMdξ_t, d2Mdξ2_t}
+    ip::IP             # ::Interpolation                Geometric interpolation 
+    M::M_t             # ::AbstractVector{<:Number}     Values of geometric shape functions
+    dMdξ::dMdξ_t       # ::AbstractVector{<:Vec}        Gradients of geometric shape functions in ref-domain
+    d2Mdξ2::d2Mdξ2_t   # ::AbstractVector{<:Tensor{2}}  Hessians of geometric shape functions in ref-domain
+                       # ::Nothing                      When hessians are not required
+end
+function GeometryMapping(::Type{T}, ip::ScalarInterpolation, qr::QuadratureRule, ::RequiresHessian{RH}) where {T,RH}
+    n_shape = getnbasefunctions(ip)
+    n_qpoints = getnquadpoints(qr)
+    VT   = Vec{getdim(ip),T}
+    M    = zeros(T,  n_shape, n_qpoints)
+    dMdξ = zeros(VT, n_shape, n_qpoints)
+    
+    # Hessian (if needed, else nothing)
+    HT = Tensor{2,getdim(ip),T}
+    d2Mdξ2 = RH ? zeros(HT, n_shape, n_qpoints) : nothing
+
+    geo_mapping = GeometryMapping(ip, M, dMdξ, d2Mdξ2)
+    precompute_values!(geo_mapping, qr)
+    return geo_mapping
+end
+function Base.copy(v::GeometryMapping)
+    d2Mdξ2_copy = v.d2Mdξ2 === nothing ? nothing : copy(v.d2Mdξ2)
+    return GeometryMapping(copy(v.ip), copy(v.M), copy(v.dMdξ), d2Mdξ2_copy)
+end
+
+precompute_values!(gm::GeometryMapping, qr::QuadratureRule) = precompute_values!(gm, RequiresHessian(gm), qr)
+
+function precompute_values!(gm::GeometryMapping, ::RequiresHessian{false}, qr::QuadratureRule)
```

Comment:
Yes, I thought of that but wasn't sure. What made me decide against is that `shape_gradients_and_values!` and siblings are likely to become public for the GPU support, so wasn't sure if it is a good idea to add internal methods to this. 

In that case, I think it should be `shape_xxx!(gm.M, gm.dMdxi, ..., ip_geo, qr)` such that it can be reused for both function values and geometry mapping so that it isn't connected to the `FunctionValues` and `GeometryMapping` directly. What do you think?